applications/nrf5340_audio/src/audio/le_audio_rx.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <nrfx_clock.h>
+
+#include "nrf5340_audio_common.h"
+#include "streamctrl.h"
+#include "audio_datapath.h"
+#include "macros_common.h"
+#include "audio_system.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(le_audio_rx, CONFIG_AUDIO_DATAPATH_LOG_LEVEL);
+
+static bool initialized;
+static struct k_thread audio_datapath_thread_data;
+static k_tid_t audio_datapath_thread_id;
+K_THREAD_STACK_DEFINE(audio_datapath_thread_stack, CONFIG_AUDIO_DATAPATH_STACK_SIZE);
+
+struct ble_iso_data {
+	uint8_t data[CONFIG_BT_ISO_RX_MTU];
+	size_t data_size;
+	bool bad_frame;
+	uint32_t sdu_ref;
+	uint32_t recv_frame_ts;
+} __packed;
+
+struct rx_stats {
+	uint32_t recv_cnt;
+	uint32_t bad_frame_cnt;
+	uint32_t data_size_mismatch_cnt;
+};
+
+DATA_FIFO_DEFINE(ble_fifo_rx, CONFIG_BUF_BLE_RX_PACKET_NUM, WB_UP(sizeof(struct ble_iso_data)));
+
+/* Callback for handling ISO RX */
+void le_audio_rx_data_handler(uint8_t const *const p_data, size_t data_size, bool bad_frame,
+			      uint32_t sdu_ref, enum audio_channel channel_index,
+			      size_t desired_data_size)
+{
+	int ret;
+	uint32_t blocks_alloced_num, blocks_locked_num;
+	struct ble_iso_data *iso_received = NULL;
+
+	if (!initialized) {
+		ERR_CHK_MSG(-EPERM, "Data received but le_audio_rx is not initialized");
+	}
+
+	/* Capture timestamp of when audio frame is received */
+	uint32_t recv_frame_ts = nrfx_timer_capture(&audio_sync_timer_instance,
+						    AUDIO_SYNC_TIMER_CURR_TIME_CAPTURE_CHANNEL);
+
+	/* Since the audio datapath thread is preemptive, no actions on the
+	 * FIFO can happen whilst in this handler.
+	 */
```

Comment:
This comment should either be remove, or there should be some mechanism here to make sure the FIFOs are changed while in this handler